### PR TITLE
Fix hybrid composition bugs

### PR DESCRIPTION
**Pull request by blasten**
## Description

Fixes bugs in the hybrid mode. For example:

1. Currently, we destroy the `SurfaceView`, which causes `PlatformView::NotifyDestroyed()` and therefore `Shell::OnPlatformViewDestroyed()`.

2. We remove the  `SurfaceView` from the view hierarchy _too_ early. Instead, we will keep in the background and rendered while we get the new one ready. This avoids an intermediary flash when switching the surface.

3. The Rtree and the picture recorders are never updated.

4. Dynamic surface switch should occur in `onDisplayPlatformView` as there could be platform views, but no overlay surfaces.

## Related Issues

https://github.com/flutter/flutter/issues/55270
https://github.com/flutter/flutter/issues/58291

## Tests

I added the following tests: These composite case are covered by the scenario app, but the presubmits aren't running.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java [hunk 1]
```diff
@@ -557,17 +557,20 @@ private void flushAllViews() {
...
-    FlutterView flutterView = (FlutterView) this.flutterView;
+  private void initializeRootImageViewIfNeeded() {
     if (!flutterViewConvertedToImageView) {
       flutterView.convertToImageView();
       flutterViewConvertedToImageView = true;
     }
+  }
 
+  public void onDisplayPlatformView(int viewId, int x, int y, int width, int height) {
+    initializeRootImageViewIfNeeded();
```
Comment (blasten) [inline shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:568]: Switches to `FlutterImageView`.

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java [hunk 2]
```diff
@@ -557,17 +557,20 @@ private void flushAllViews() {
...
+    // TODO: Implement this method. https://github.com/flutter/flutter/issues/58288
+  }
+
+  public void onDisplayOverlaySurface(int id, int x, int y, int width, int height) {
+    initializeRootImageViewIfNeeded();
```
Comment (blasten) [inline shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:573]: This one is redundant, perhaps it can be removed, but it doesn't hurt to do it in both places.

File: shell/platform/android/platform_view_android.cc [hunk 3]
```diff
@@ -104,6 +104,22 @@ void PlatformViewAndroid::NotifyCreated(
...
+void PlatformViewAndroid::NotifySurfaceWindowChanged(
+    fml::RefPtr<AndroidNativeWindow> native_window) {
+  if (android_surface_) {
+    fml::AutoResetWaitableEvent latch;
+    fml::TaskRunner::RunNowOrPostTask(
+        task_runners_.GetRasterTaskRunner(),
+        [&latch, surface = android_surface_.get(),
+         native_window = std::move(native_window)]() {
+          surface->TeardownOnScreenContext();
+          surface->SetNativeWindow(native_window);
```
Comment (blasten) [inline shell/platform/android/platform_view_android.cc:116]: this new native API just tears down the on screen context and sets the new native window.

File: shell/platform/android/io/flutter/embedding/android/FlutterImageView.java [hunk 4]
```diff
@@ -62,29 +82,56 @@ public FlutterRenderer getAttachedRenderer() {
...
-    nextImage = imageReader.acquireLatestImage();
+    // There's no guarantee that the image will be closed before the next call to
+    // `acquireLatestImage()`. For example, the device may not produce new frames if
+    // it's in sleep mode, so the calls to `invalidate()` will be queued up
+    // until the device produces a new frame.
+    //
+    // While the engine will also stop producing frames, there is a race condition.
+    //
+    // To avoid exceptions, check if a new image can be acquired.
+    if (pendingImages < imageReader.getMaxImages()) {
```
Comment (blasten) [inline shell/platform/android/io/flutter/embedding/android/FlutterImageView.java:129]: To repro this bug: while running the app, turn off the screen and the app crashes

File: shell/platform/android/io/flutter/embedding/android/FlutterImageView.java [hunk 5]
```diff
@@ -41,13 +41,33 @@
   @Nullable private Bitmap currentBitmap;
   @Nullable private FlutterRenderer flutterRenderer;
 
+  public enum SurfaceKind {
+    /** Displays the background canvas. */
+    background,
+
+    /** Displays the overlay surface canvas. */
+    overlay,
+  }
```
Comment (blasten) [inline shell/platform/android/io/flutter/embedding/android/FlutterImageView.java:50]: The java/c++ flows are very different depending on the surface kind. Maybe, the code can be refactored to have a unified flow.  This is out of scope at the moment though.

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java [hunk 6]
```diff
@@ -54,7 +54,7 @@
   private Context context;
 
   // The View currently rendering the Flutter UI associated with these platform views.
-  private View flutterView;
+  private FlutterView flutterView;
```
Comment (jason-simmons) [inline shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:57]: Does `PlatformViewsController` still need to support the legacy `io.flutter.view.FlutterView`?

File: shell/platform/android/io/flutter/embedding/android/FlutterImageView.java (hunk 4 above)
Comment (jason-simmons) [reply shell/platform/android/io/flutter/embedding/android/FlutterImageView.java:129]: Does the error still happen if you increase maxImages as done in https://github.com/flutter/engine/pull/19319?

See Android's recommendation in https://developer.android.com/reference/android/media/ImageReader#acquireLatestImage()

File: shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java [hunk 7]
```diff
@@ -375,7 +375,11 @@ public void systemInsetHandlesFullscreenNavbarLeft() {
   public void flutterImageView_acquiresImageAndInvalidates() {
     final ImageReader mockReader = mock(ImageReader.class);
     final FlutterImageView imageView =
-        spy(new FlutterImageView(RuntimeEnvironment.application, mockReader));
+        spy(
```
Comment (jason-simmons) [inline shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java:378]: `FlutterViewTest.flutterImageView_acquiresImageAndInvalidates` is failing on LUCI

File: shell/platform/android/platform_view_android.cc (hunk 3 above)
Comment (jason-simmons) [reply shell/platform/android/platform_view_android.cc:116]: What is the difference between this and calling `FlutterRenderer.stopRenderingToSurface`/`FlutterRenderer.startRenderingToSurface`?

File: shell/platform/android/io/flutter/embedding/android/FlutterImageView.java (hunk 4 above)
Comment (blasten) [reply shell/platform/android/io/flutter/embedding/android/FlutterImageView.java:129]: yeah. `onDraw()` it throttled or not called at all, which leaves the images open.

Comment (blasten) [reply shell/platform/android/io/flutter/embedding/android/FlutterImageView.java:129]: yeah. onDraw() is throttled or not called at all, which leaves the images open.

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java (hunk 6 above)
Comment (blasten) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:57]: I don't believe so since we migrated to the new embedding

Comment (blasten) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:57]: Actually, I will change that back and investigate if the old embedding isn't supported.

File: shell/platform/android/platform_view_android.cc (hunk 3 above)
Comment (blasten) [reply shell/platform/android/platform_view_android.cc:116]: `FlutterRenderer.stopRenderingToSurface()` calls `PlatformView::NotifyDestroyed()`, and then `Shell::OnPlatformViewDestroyed()`, which tears down the rasterizer: https://github.com/flutter/engine/blob/35d7d4aa6cf9560bf691f383c0a65cf982b65004/shell/common/shell.cc#L723 

It also causes [this issue](https://github.com/flutter/engine/blob/master/shell/gpu/gpu_surface_gl.cc#L215), which is due to [this method](https://github.com/google/skia/blob/c19b9c58862feee3a482575ae014293da02bda43/src/gpu/GrProxyProvider.cpp#L889) returning `true`. I asked about it on the Flutter & Skia chat, and @chinmaygarde had some context about previous cases.

 I haven't figured out exactly why the sequence `FlutterRenderer.stopRenderingToSurface`/`FlutterRenderer.startRenderingToSurface` is very different than just tearing down the `OnScreenContext` and then setting the native window.

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java (hunk 6 above)
Comment (blasten) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java]: Done.

File: shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java (hunk 7 above)
Comment (blasten) [reply shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java:382]: Fixed.

File: shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java [hunk 8]
```diff
@@ -386,6 +390,29 @@ public void flutterImageView_acquiresImageAndInvalidates() {
...
+    when(mockReader.acquireLatestImage()).thenReturn(mockImage);
+
+    final FlutterImageView imageView =
+        spy(
+            new FlutterImageView(
+                RuntimeEnvironment.application,
+                mockReader,
+                FlutterImageView.SurfaceKind.background));
+
+    doNothing().when(imageView).invalidate();
```
Comment (blasten) [inline shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java:408]: Added this test case. `invalidate()` doesn't provide guarantees on when `onDraw(Canvas canvas)` is called.

https://developer.android.com/reference/android/view/View#invalidate()
> Invalidate the whole view. If the view is visible, onDraw(android.graphics.Canvas) will be called at some point in the future.

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java (hunk 6 above)
Comment (blasten) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java]: @xster do you know if the old Android embedding is not longer supported?

File: shell/platform/android/io/flutter/embedding/android/FlutterImageView.java (hunk 4 above)
Comment (mutant0113) [reply shell/platform/android/io/flutter/embedding/android/FlutterImageView.java:129]: Hi, I faced this crash issue recently. I'd like to ask do we have any solution for this?

Comment (mutant0113) [reply shell/platform/android/io/flutter/embedding/android/FlutterImageView.java:129]: Please refer to this issue link https://github.com/flutter/flutter/issues/63897. Thanks.
